demo/lightning/checkpoint/multinode/fsspecfsdp.py

```diff
@@ -0,0 +1,107 @@
+import os
+
+import gcsfs
+import torch
+
+from pathlib import Path
+from typing import Generator
+from lightning.pytorch.trainer.states import TrainerFn
+from lightning.pytorch.strategies import FSDPStrategy
+from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
+
+from lightning.pytorch.strategies.fsdp import _METADATA_FILENAME
+from torch.distributed.checkpoint import save, load
+from torch.nn import Module
+from torch.distributed.checkpoint import _fsspec_filesystem as FF
+
+
+class FSSpecFSDPStrategy(FSDPStrategy):
+
+    def __init__(self, path, model, **kwargs):
+        super().__init__(**kwargs)
+        self.model = model
+        self.path = path
+
+    def save_checkpoint(self,
+                        checkpoint,
+                        filepath,
+                        storage_options=None) -> None:
+        if storage_options is not None:
+            raise TypeError(
+                "`FSDPStrategy.save_checkpoint(..., storage_options=...)` is not supported because"
+                " `FSDPStrategy` does not use the `CheckpointIO`.")
+
+        writer = FF.FsspecWriter(self.path, sync_files=False)
+        _ = Path(self.broadcast(filepath))
+
+        converted_state = {"model": checkpoint.pop("state_dict")}
+        converted_state.update({
+            f"optimizer_{idx}": optim_state
+            for idx, optim_state in enumerate(
+                checkpoint.pop("optimizer_states", []))
+        })
+        save(converted_state, checkpoint_id=filepath, storage_writer=writer)
+
+        bucket = gcsfs.GCSFileSystem()
+        with bucket.open(os.path.join(filepath, _METADATA_FILENAME),
+                         'wb') as f:
+            torch.save(checkpoint, f)
+
+    def get_sharded_state_dict_context(
+            self, module: Module) -> Generator[None, None, None]:
+
+        from torch.distributed.fsdp.api import ShardedOptimStateDictConfig, ShardedStateDictConfig, StateDictType
+
+        state_dict_config = ShardedStateDictConfig(offload_to_cpu=True)
+        optim_state_dict_config = ShardedOptimStateDictConfig(
+            offload_to_cpu=True)
+        state_dict_type_context = FSDP.state_dict_type(
+            module=module,
+            state_dict_type=StateDictType.SHARDED_STATE_DICT,
+            state_dict_config=state_dict_config,
+            optim_state_dict_config=optim_state_dict_config,
+        )
+        return state_dict_type_context  # type: ignore[return-value]
+
+    def load_checkpoint(self, checkpoint_path):
+        # broadcast the path from rank 0 to ensure all the states are loaded from a common path
+        _ = Path(self.broadcast(checkpoint_path))
```

Comment:
Same question about the empty path, this looks like its used directly [in the normal fsdp implementation](https://github.com/Lightning-AI/pytorch-lightning/blob/master/src/lightning/pytorch/strategies/fsdp.py#L591).

Reply:
Path() converts the path passed to it to a posix path. For example, Path("gs://checkpoint-benchmark-west1") would return "gs:/checkpoint-benchmark-west1" which would break things when fsspec tries to read/write to/from the bucket (but OK for the normal FSDP because it assumes local paths). But I think we would still need to broadcast the patch to all nodes. 

It's probably okay to just do `self.broadcast(path)`.

Reply:
Ok, is this broadcast operation meant to be done on all nodes or just the rank 0 node?